### ISSUE TITLE
Comments on proposed diagnostics flowchart

### ISSUE THREAD

**Issue by dangunter** [IDAES/idaes-pse#1208]
See https://github.com/IDAES/idaes-pse/wiki/Diagnostics-Flowchart

The flowchart image is stored at [docs/static/extra/DiagnosticsFlowchart.png](https://github.com/IDAES/idaes-pse/blob/main/docs/static/extra/DiagnosticsFlowchart.png)

- [ ] Flowchart needs step to create instance of Diagnostics class

**Comment by mrmundt** [comment]
General Comments:
- "Your feedback is needed! Please look at this flowchart and explanation (below it) and use [this issue (#1208)](https://github.com/IDAES/idaes-pse/issues/1208) to add questions and comments." <- There is no Explanation.
- The color coding needs a legend. Why is one box yellow vs. blue?
- The different shapes also need a legend. Does rectangle/sharp-edges mean something different from rectangle/round-edges? If the edge shape doesn't matter, be consistent.
- There also needs to be a legend for acronyms. I have assumptions that are probably correct, but we never want to make assumptions about what our users know, if possible
- Some of the line colors aren't consistent (e.g., line coming out of "Issue Found?" should probably be orange)

Specific Comments:

(_NOTE_: I am not an IDAES-pse user, so I might not know certain terminology)

-  Is a user expected to know what a Logbook is?
- What do you mean by "Issues"? Will there be a page to detail what "Issues" might look like?
- Do you really need to call out "Look at logs"? I don't know if the average user would be inclined to do that anyway
- If a solution isn't returned, how do you get a partial solution?
- Is "Poor Behavior" quantifiable in some way? What does that mean?
- Does a person know how to write Robustness tests? Will there be examples for that?
- The whole furthest left-hand side of the workflow is really hard to follow. I am not sure what the arrows even mean or are trying to imply for some of them.
- What is meant by "Clean solve"?

**Comment by andrewlee94** [comment]
@mrmundt Regarding your comment about the left-hand side, I am assuming the issue is that everything basically loops back to the "Revise Model" box at the top of the flowchart (which is off-screen by the time you get even part way down). Is this correct, and if so do you have any suggestions on how to make that clearer?

**Comment by mrmundt** [comment]
> @mrmundt Regarding your comment about the left-hand side, I am assuming the issue is that everything basically loops back to the "Revise Model" box at the top of the flowchart (which is off-screen by the time you get even part way down). Is this correct, and if so do you have any suggestions on how to make that clearer?

It's partially that and partially that, because the "arrow-line" colors change, I am unsure where they are supposed to be leading. There is one instance, for example, where there is a "yes" line that meets up with another line that I don't think is also a yes line? But they go to the same place.

![Screen Shot 2023-06-15 at 14 54 25](https://github.com/IDAES/idaes-pse/assets/55767766/24874006-e2f6-4223-97af-00c202bcde6e)

There are so many degrees on that line that I have NO Idea where they are supposed to go or what that line is supposed to mean anymore.

**Comment by andrewlee94** [comment]
To start with, I am thinking of dropping the coloring of the lines (in part for this reason). Do you think it would also help to have arrows at each junction showing all the "inflows" so the "outflow" is clear(er).

**Comment by mrmundt** [comment]
> To start with, I am thinking of dropping the coloring of the lines (in part for this reason). Do you think it would also help to have arrows at each junction showing all the "inflows" so the "outflow" is clear(er).

I think both of those things would help, yes! Anything that will help the user from having to trace lines on their screen with their fingers ;)

**Comment by andrewlee94** [comment]
@mrmundt I've (or more accurately @dangunter) uploaded a new version of the flowchart and added some initial explanation of the steps. If you have some time, w would appreciate more feedback.

**Comment by MarcusHolly** [comment]
A few small comments:

1.) Does the `Diagnostics Toolbox` currently exist? If so, where is it located?

2.) What specific kind of issues would Step 5 (Run Structural Analysis) resolve? I'm having a hard time imagining what a "model structure" issue would look like, so providing an example/more explanation would be helpful. Is it primarily DOF-related issues? Perhaps this is/would be explained in a docstring or documentation

**Comment by andrewlee94** [comment]
@MarcusHolly

1. No, the toolbox does not exist yet but we are starting on the prototype.
2. Structural analysis is more than just degrees of freedom. We can also look for over and under constrained parts of the problem (structural singularities), potential evaluation errors (e.g. places where you can get division by zero), unused variables, variables fixed to a hard zero, etc. And yes, there will eventually be more documentation to go with this.

**Comment by avdudchenko** [comment]
Thanks for putting this together. I think this workflow makes sense for unit-level model development. I support testing optimizations/solving over broad solution space for a unit model. This will help ensure good model behavior, especially in a flowsheet.  

Some comments:
1) I assume this workflow and toolbox are really designed for unit-level analysis and would not quite work on a flowsheet level (e.g. multiple units together?) (e.g. if  I am using out-of-the-box unit models, I don't quite see how I would find a way to revise the unit models them-self?) 

2) Are there any quantitative metrics/scores that would define a given model as "Robust" or having "good behavior" (some times you can get a model to solve, but it takes 5000 iterations....is that a good behaving model?) (This is same as @mrmundt comment) It would be good to provide devs. with what some quantitative targets and include those in tests? Not sure how possible this is.

3) Regarding the "Revise model,". its not quiet clear if the logs would provide guidance on how to do this or where to look? Is there going to be a set of instructions that helps develop better model formulation?  (e.g. there might be some obvious issues, divisions by zero, log type formulation, and some not-so-obvious ones?). I think we would need another flow chart for going through the model revision process and accompanying doc.

**Comment by andrewlee94** [comment]
@avdudchenko Some response to your comments:

1. Yes and no. The tools, methods and workflow will work for both flowsheets and unit model, and will be useful for both. However, as you noted, a flowsheet developer will often not know enough about a unit model to be able to fix any issues that are identified. In the end, we are hoping that this will be a tiered process - the unit model developer will test their model and make sure there are no issues so that when a flowsheet developer comes to use it the only issues they will see are issues that arise from the flowsheet itself. To help with this, the tool are also going to include testing methods that will hopefully become part of the model testing framework so that we can ensure that unit models are developed to be robust.
2. The first and most important metric will be that the tools do not identifying any critical issues - these will be independent of the solver so this can be a concrete check. As far as solver metrics go, that is somewhat more open ended and probably not something we will get into just yet (plus it will depend on the solver used).
3. You are correct- finding the issues is only the first part of the problem and fixing them is a whole different issue. At this point we are focusing on the first part; helping users identify what issues they have. Fixing the issues is always going to be more of an art than a science and will require knowledge of the model being worked on so it is hard to write anything general about that.

**Comment by andrewlee94** [comment]
@mrmundt Thank you for the comments - they will be important when we start writing up the actual documentation for this.

We have been discussing many of your points as we work on this. At this point, we are thinking there will be a series of documents, of which this workflow will probably be the last as the "overall master workflow for experience users"; we do recognize that there needs to be a lot of introductory material before this. However, we are also knowing on ideas about making the tools self-guiding, and having them suggest the next steps to the user so that this workflow would become a reference more than a guide.

**Comment by andrewlee94** [comment]
Closing as it seems to have run its course.
